### PR TITLE
Add a lora dense layer

### PR DESCRIPTION
**Pull request by mattdangerw**
https://github.com/keras-team/keras-nlp/pull/1264 shows how this will eventually fit together.

### CODE REVIEW

**Comment by mattdangerw** [comment]
Note that this is Keras 3/Keras Core only because that library will allow you to set trainable on individual parameters (tf.keras does not). It doesn't seem worth the effort to build backwards compat here, this can be a strictly forward looking Keras 3 API.

File: keras_nlp/layers/modeling/lora_dense.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+            f"Received invalid equation `'{equation}'`."
+        )
+
+
+@keras_nlp_export("keras_nlp.layers.LoraDense")
+class LoraDense(keras.layers.Layer):
+    """A LoRA adapter layer for a dense input layer.
+
+    This layer implements a low-rank decomposition of a dense transformation, as
+    described in [LoRA: Low-Rand Adaptation Of Large Language Models](https://arxiv.org/pdf/2106.09685.pdf)
```
Comment (ianstenbit) [inline keras_nlp/layers/modeling/lora_dense.py:43]: nit: Low-Rank, not Low-Rand

File: keras_nlp/layers/modeling/lora_dense.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+    This layer can be used to replace a dense layer with a layer who's
```
Comment (ianstenbit) [inline keras_nlp/layers/modeling/lora_dense.py:44]: whose

File: keras_nlp/layers/modeling/lora_dense.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    parameters are mostly frozen.
+
+    By default, this layer takes in a `inner_dense` layer, freezes it's
```
Comment (ianstenbit) [inline keras_nlp/layers/modeling/lora_dense.py:47]: its

File: keras_nlp/layers/modeling/lora_dense.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    `inner_dense` kernel by calling `merge_weights()`.
+
+    Args:
+        inner_dense: A `keras.layers.Dense` or `keras.layers.EinsumDense`.
+            The inner dense layer to freeze and wrap with the `LoraDense`
+            layer. Note that for `EinsumDense` layers, the einsum equation must
+            represent a dense transformation on the last axis of the input,
+            though adding new axes to the output (e.g. a multi-head axis) is
+            allowed.
+        rank: int The inner rank of the decomposed dense transformation. The
```
Comment (ianstenbit) [inline keras_nlp/layers/modeling/lora_dense.py:59]: Looks like for consistency, we want a period after "int"

File: keras_nlp/layers/modeling/lora_dense.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+            lower this number, the less trainable parameters the layer will
```
Comment (ianstenbit) [inline keras_nlp/layers/modeling/lora_dense.py:60]: s/less/fewer

File: keras_nlp/layers/modeling/lora_dense.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+            "When passing a `EinsumDense` layer to a `LoraDense` layer, the "
+            "einsum `equation` must always have the form `*x,x*->*`, where "
+            "each `*` can be any sequence. Conceptually, the `equation` should "
+            "always represent a dense matmul on the last axis of the input. "
+            f"Received invalid equation `'{equation}'`."
+        )
+
+
+@keras_nlp_export("keras_nlp.layers.LoraDense")
+class LoraDense(keras.layers.Layer):
```
Comment (ianstenbit) [inline keras_nlp/layers/modeling/lora_dense.py:39]: I'm not opinionated, just want to make sure it's considered -- should this ever be LoRADense? IIRC the Keras convention is to exclude this sort of capitalization but I can't remember off the top of my head.

File: keras_nlp/layers/modeling/lora_dense.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+    lora_dense(inputs)
+    ```
+    """
+
+    def __init__(
+        self,
+        inner_dense,
+        rank=8,
+        alpha=32.0,
+        lora_a_initializer="variance_scaling",
```
Comment (ianstenbit) [inline keras_nlp/layers/modeling/lora_dense.py:116]: Does it make sense to include "lora" as a prefix for this variable name, given that it's already a LoraDense layer/

File: keras_nlp/layers/modeling/lora_dense_test.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        input_data = ops.random.uniform((2, 4, 16))
+        lora_output = layer(input_data)
+        dense_output = inner_dense(input_data)
+        self.assertNotAllClose(lora_output, dense_output)
+
+        layer.merge_weights()
+        merged_lora_output = layer(input_data)
+        dense_output = inner_dense(input_data)
+        self.assertAllClose(lora_output, merged_lora_output)
+        self.assertAllClose(lora_output, dense_output)
```
Comment (ianstenbit) [inline keras_nlp/layers/modeling/lora_dense_test.py:97]: It might be worthwhile to include a test case about the trainability of all the weights included in the layer, especially since these "freeze" params exist and they're acted upon in `build`

**Review by ianstenbit — CHANGES_REQUESTED** [review]

File: keras_nlp/layers/modeling/lora_dense.py [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+    # Output shape (4, 10, 2, 8)
+    lora_dense(inputs)
+    ```
+    """
+
+    def __init__(
+        self,
+        inner_dense,
+        rank=8,
+        alpha=32.0,
```
Comment (abheesht17) [inline keras_nlp/layers/modeling/lora_dense.py:115]: Generally, `alpha = rank`. I know the guide has `alpha = 32.`, and `rank = 4`, but it was an oversight on my part. The authors state that they went with `alpha = rank`, and tuned the learning rate.

Source: Section 4.1 in https://openreview.net/forum?id=nZeVKeeFYf9

File: keras_nlp/layers/modeling/lora_dense.py [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+            lower this number, the less trainable parameters the layer will
+            have.
+        alpha: float. A constant value used for scaling the lora update. The
+            lora update to the original dense transformation will be scaled by
+            `alpha / rank`.
+        lora_a_initializer: The initializer to use for the inner projection
+            from layer inputs to the inner `rank` intermediate outputs.
+        freeze_kernel: If true, the kernel of the inner dense layer will have
+            `trainable` set to False.
+        freeze_bais: If true, the kernel of the inner dense layer will have
```
Comment (abheesht17) [inline keras_nlp/layers/modeling/lora_dense.py:69]: bais --> bias

File: keras_nlp/layers/modeling/lora_dense.py (hunk 3 above)
Comment (abheesht17) [reply keras_nlp/layers/modeling/lora_dense.py:47]: in a `inner_dense` layer --> in an `inner_dense` layer?

File: keras_nlp/layers/modeling/lora_dense.py (hunk 7 above)
Comment (abheesht17) [reply keras_nlp/layers/modeling/lora_dense.py:116]: I think it's okay, because the two dense layers in LoRA are called "lora_A" and "lora_B" (in the [official code](https://github.com/microsoft/LoRA/blob/main/loralib/layers.py#L48)). Calling this `a_initializer` would look weird :P.

**Review by abheesht17 — APPROVED** [review]
Was passing through, left some NITs

File: keras_nlp/layers/modeling/lora_dense.py (hunk 6 above)
Comment (mattdangerw) [reply keras_nlp/layers/modeling/lora_dense.py:39]: Yeah, when I discussed this with Francois we landed on if it is pronounceable and colloquially said, consider it a single segment in our camel casing. So "Roberta" is just that, not RoBERTa. Honestly it would be a bit of a homework assignment to remeber where all the casing goes.

We actually have a line for this in our style guide! https://github.com/keras-team/keras-nlp/blob/master/STYLE_GUIDE.md#naming-of-layers-and-models

File: keras_nlp/layers/modeling/lora_dense.py (hunk 7 above)
Comment (mattdangerw) [reply keras_nlp/layers/modeling/lora_dense.py:116]: Yeah, I wanted to avoid `layer.a` and `layer.b`. We could come up with our own names, `layer.inner_kernel_update` and `layer.outer_kernel_update`, but I suspect `lora_a` and `lora_b` will be more recognizable to people.

File: keras_nlp/layers/modeling/lora_dense.py (hunk 9 above)
Comment (mattdangerw) [reply keras_nlp/layers/modeling/lora_dense.py:115]: Thanks! Seems like we could

- Default `alpha=8.`, so it's obvious the type.
- Default `alpha=None` and assign `float(rank)` if unset.
- Just expose `scale` directly, and default `scale=1.`.
- Don't expose anything.

Maybe just `alpha=8.`? Simple and what people would expect?

File: keras_nlp/conftest.py [hunk 11]
```diff
@@ -108,10 +108,14 @@ def pytest_collection_modifyitems(config, items):
         not run_tpu,
         reason="need --run_tpu option to run",
     )
-    skip_tf_only = pytest.mark.skipif(
+    tf_only = pytest.mark.skipif(
         not backend_config.backend() == "tensorflow",
         reason="tests only run on tf backend",
     )
+    multi_backend_only = pytest.mark.skipif(
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_nlp/conftest.py:115]: hmm on second thought i'll leave it as multi-backend for now. because it can run with keras 3 or keras core for the time being. and we will have zero coverage for the actual keras 3 code path till we get that pip.

**Comment by mattdangerw** [comment]
I think this is ready for another round!

File: keras_nlp/layers/modeling/lora_dense.py (hunk 9 above)
Comment (ianstenbit) [reply keras_nlp/layers/modeling/lora_dense.py]: I think this approach LG

File: keras_nlp/layers/modeling/lora_dense.py [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+        rank: int. The inner rank of the decomposed dense transformation. The
+            lower this number, the fewer trainable parameters the layer will
+            have.
+        alpha: float. A constant value used for scaling the lora update. The
+            lora update to the original dense transformation will be scaled by
+            `alpha / rank`.
+        lora_a_initializer: The initializer to use for the inner projection
+            from layer inputs to the inner `rank` intermediate outputs.
+        freeze_kernel: If true, the kernel of the inner dense layer will have
+            `trainable` set to False.
```
Comment (ianstenbit) [inline keras_nlp/layers/modeling/lora_dense.py:72]: IIRC we backtick `False` in these contexts?

**Review by ianstenbit — APPROVED** [review]
Looks good!

File: keras_nlp/layers/modeling/lora_dense.py (hunk 7 above)
Comment (fchollet) [reply keras_nlp/layers/modeling/lora_dense.py:120]: Sounds fine to me.
